packages/core/contracts/oracle/implementation/SkinnyOptimisticOracle.sol

```diff
@@ -0,0 +1,716 @@
+        uint256 _reward,
+        uint256 _bond,
+        uint256 _customLiveness
+    ) external override nonReentrant() returns (uint256 totalBond) {
+        bytes32 requestId = _getId(msg.sender, _identifier, _timestamp, _ancillaryData);
+        require(requests[requestId] == bytes32(0), "Request already initialized");
+        require(_getIdentifierWhitelist().isIdentifierSupported(_identifier), "Unsupported identifier");
+        require(_getCollateralWhitelist().isOnWhitelist(address(_currency)), "Unsupported currency");
+        require(_timestamp <= getCurrentTime(), "Timestamp in future");
+        require(
+            _stampAncillaryData(_ancillaryData, msg.sender).length <= ancillaryBytesLimit,
+            "Ancillary Data too long"
+        );
+        uint256 finalFee = _getStore().computeFinalFee(address(_currency)).rawValue;
+
+        // Associate new request with ID
+        Request memory request;
+        request.currency = _currency;
+        request.reward = _reward;
+        request.finalFee = finalFee;
+        request.bond = _bond != 0 ? _bond : finalFee;
+        request.customLiveness = _customLiveness;
+
+        if (_reward > 0) {
+            _currency.safeTransferFrom(msg.sender, address(this), _reward);
+        }
+
+        _storeRequestHash(requestId, request);
+        emit RequestPrice(msg.sender, _identifier, _timestamp, _ancillaryData, request);
+
+        return request.bond.add(request.finalFee);
+    }
+
+    /**
+     * @notice Proposes a price value on another address' behalf. Note: this address will receive any rewards that come
+     * from this proposal. However, any bonds are pulled from the caller.
+     * @param _requester sender of the initial price request.
+     * @param _identifier price identifier to identify the existing request.
+     * @param _timestamp timestamp to identify the existing request.
+     * @param _ancillaryData ancillary data of the price being requested.
+     * @param _request price request parameters whose hash must match the request that the caller wants to
+     * propose a price for.
+     * @param _proposer address to set as the proposer.
+     * @param _proposedPrice price being proposed.
+     * @return totalBond the amount that's pulled from the caller's wallet as a bond. The bond will be returned to
+     * the proposer once settled if the proposal is correct.
+     */
+    function proposePriceFor(
+        address _requester,
+        bytes32 _identifier,
+        uint256 _timestamp,
+        bytes memory _ancillaryData,
+        Request memory _request,
+        address _proposer,
+        int256 _proposedPrice
+    ) public override nonReentrant() returns (uint256 totalBond) {
+        require(_proposer != address(0), "proposer address must be non 0");
+        require(
+            _getState(_requester, _identifier, _timestamp, _ancillaryData, _request) ==
+                OptimisticOracleInterface.State.Requested,
+            "proposePriceFor: Requested"
```

Comment:
this is a different `require` style to some of the other methods. should we convert on something more consistant?

Reply:
Yeah do you prefer having the function name in the error message? I see this sometimes used in the OO @mrice32

Reply:
Tbh, it's not a huge deal either way. We just want the error messages to be as unique as possible so a caller can find the line of code where it was emitted.